### PR TITLE
Fix DiscoverModelMetadata throw on QA models with dynamic output dims

### LINKED ISSUE #25: Bug: DiscoverModelMetadata throws on QA models with 2D dynamic output tensors (start_logits/end_logits)
### Summary

`OnnxTextModelScorerEstimator.DiscoverModelMetadata()` unconditionally throws `InvalidOperationException("Could not determine hidden dimension from ONNX output 'start_logits'")` when used with Question Answering models. The root cause is that the primary output tensor's dimension-extraction logic assumes at least one static (non-dynamic) dimension, but QA model outputs (`start_logits`, `end_logits`) have shape `[batch_size, sequence_length]` where **both** dimensions are dynamic (`[-1, -1]`).

**Ironically, the adjacent code for `AdditionalOutputTensorNames` already handles this correctly** by falling back to `MaxTokenLength` when the last dimension is dynamic. The fix is to apply the same fallback to the primary output.

### Error

```
System.InvalidOperationException: Could not determine hidden dimension from ONNX output 'start_logits'.
   at MLNet.TextInference.Onnx.OnnxTextModelScorerEstimator.DiscoverModelMetadata(InferenceSession session)
   at MLNet.TextInference.Onnx.OnnxTextModelScorerEstimator.Fit(IDataView input)
   at MLNet.TextInference.Onnx.OnnxQaEstimator.Fit(IDataView input)
```

### Reproduction

```csharp
var mlContext = new MLContext();
var qaOptions = new OnnxQaOptions
{
    ModelPath = "/path/to/minilm-uncased-squad2/model.onnx",
    TokenizerPath = "/path/to/minilm-uncased-squad2/vocab.txt",
};
var estimator = new OnnxQaEstimator(mlContext, qaOptions);
var transformer = estimator.Fit(inputDataView); //  throws here
```

Tested with `lquint/minilm-uncased-squad2-onnx` model (from PR #24).

### ONNX Model Tensor Shapes

Verified with `onnxruntime` Python inspection:

```
=== INPUTS ===
  input_ids:      shape=['batch_size', 'sequence_length'], type=tensor(int64)
  attention_mask: shape=['batch_size', 'sequence_length'], type=tensor(int64)
  token_type_ids: shape=['batch_size', 'sequence_length'], type=tensor(int64)

=== OUTPUTS ===
  start_logits:   shape=['batch_size', 'sequence_length'], type=tensor(float)
  end_logits:     shape=['batch_size', 'sequence_length'], type=tensor(float)
```

Both output dimensions are dynamic (ONNX reports them as `-1`). Compare with an embedding model like `BAAI/bge-small-en-v1.5`:

```
=== OUTPUTS ===
  last_hidden_state: shape=['batch_size', 'sequence_length', 384], type=tensor(float)
```

Embedding models have a **static** last dimension (384 = hidden_dim), but QA models have **all dynamic** dimensions.

### Root Cause Analysis

**File:** `src/MLNet.TextInference.Onnx/OnnxTextModelScorerEstimator.cs`
**Method:** `DiscoverModelMetadata(InferenceSession session)` (~line 170)

#### Call chain

1. `OnnxQaEstimator.Fit()` creates `OnnxTextModelScorerOptions` with:
   ```csharp
   PreferredOutputNames = ["start_logits"],
   AdditionalOutputTensorNames = ["end_logits"],
   AdditionalOutputColumnNames = ["EndLogits"],
   ```

2. `DiscoverModelMetadata()` enters the `PreferredOutputNames` branch:
   ```csharp
   else if (_options.PreferredOutputNames != null)
   {
       var preferred = TryFindTensorName(outputMeta, _options.PreferredOutputNames);
       if (preferred != null) // "start_logits" IS found
       {
           outputName = preferred;                                      // "start_logits"
           var dims = outputMeta[preferred].Dimensions;                 // [-1, -1] (both dynamic)
           hasPooledOutput = !dims.Contains(-1) && dims.Length == 2;    // false
           hiddenDim = (int)dims.Last();                                //  -1 (BUG!)
           outputRank = dims.Length;                                    // 2
       }
   }
   ```

3. Then the guard clause throws:
   ```csharp
   if (hiddenDim <= 0)
       throw new InvalidOperationException(
           $"Could not determine hidden dimension from ONNX output '{outputName}'.");
   ```

#### Why the additional outputs DON'T have this bug

The additional output discovery code (same method, ~20 lines later) **already has the correct fallback**:

```csharp
var dims = outputMeta[name].Dimensions;
int lastDim = (int)dims.Last();
if (lastDim <= 0)
    additionalOutputDims[i] = _options.MaxTokenLength;  //  correct fallback!
```

The primary output is missing this same fallback pattern.

### Proposed Fix

Apply the same `MaxTokenLength` fallback to the primary output's `hiddenDim` extraction. This needs to happen in **all three** branches that compute `hiddenDim` (the `OutputTensorName` branch, the `PreferredOutputNames` branch, and the generic auto-discovery branch).

#### Minimal diff

In `OnnxTextModelScorerEstimator.cs`, `DiscoverModelMetadata()` method, change each `hiddenDim = (int)dims.Last();` line:

```diff
 // In the OutputTensorName branch:
   hiddenDim = (int)dims.Last();
+  if (hiddenDim <= 0)
+      hiddenDim = _options.MaxTokenLength;

 // In the PreferredOutputNames branch (preferred != null):
   hiddenDim = (int)dims.Last();
+  if (hiddenDim <= 0)
+      hiddenDim = _options.MaxTokenLength;

 // In the PreferredOutputNames branch (preferred == null fallback):
   hiddenDim = (int)dims.Last();
+  if (hiddenDim <= 0)
+      hiddenDim = _options.MaxTokenLength;

 // In the generic auto-discovery unpooled branch:
   hiddenDim = (int)outputMeta[outputName].Dimensions.Last();
+  if (hiddenDim <= 0)
+      hiddenDim = _options.MaxTokenLength;
```

Keep the existing guard clause as a safety net (it would only trigger if `MaxTokenLength` is also  0, which is a configuration error):

```csharp
if (hiddenDim <= 0)
    throw new InvalidOperationException(
        $"Could not determine hidden dimension from ONNX output '{outputName}'.");
```

#### Why this is correct

- For **QA models**: `start_logits` has shape `[batch, seq_len]` with dynamic `seq_len`. At runtime, `seq_len = MaxTokenLength` (128 by default). Setting `hiddenDim = MaxTokenLength` means:
  - `GetOutputSchema()` declares `outputSize = hiddenDim = 128` for the `StartLogits` column (rank 2 path) 
  - `RunOnnxBatch()` slices `b * hiddenDim` to `hiddenDim` per row (128 floats per row) 
  - `RunOnnxBatchMulti()` doesn't use `hiddenDim` at all  it dynamically computes `perRowSize = outputSpan.Length / batchSize` 

- For **embedding models**: Hidden dim is static (e.g., 384), so the fallback never triggers. No behavior change.

- For **classification/NER/reranking models**: These output 3D tensors `[batch, seq_len, hidden_dim]` where `hidden_dim` is static. No behavior change.

### Testing

After the fix, verify with:

```csharp
var mlContext = new MLContext();
var qaOptions = new OnnxQaOptions
{
    ModelPath = "model.onnx",              // from lquint/minilm-uncased-squad2-onnx
    TokenizerPath = "vocab.txt",           // from lquint/minilm-uncased-squad2-onnx
};
var estimator = new OnnxQaEstimator(mlContext, qaOptions);

var data = new[] { new { Question = "What is ML.NET?", Context = "ML.NET is a machine learning framework for .NET developers." } };
var inputDv = mlContext.Data.LoadFromEnumerable(data);
var transformer = estimator.Fit(inputDv);   // should no longer throw
var result = transformer.Transform(inputDv);

// Read QA results
using var cursor = result.GetRowCursor(result.Schema);
var answerGetter = cursor.GetGetter<ReadOnlyMemory<char>>(result.Schema["Answer"]);
var scoreGetter = cursor.GetGetter<float>(result.Schema["Score"]);

while (cursor.MoveNext())
{
    ReadOnlyMemory<char> answer = default;
    float score = 0;
    answerGetter(ref answer);
    scoreGetter(ref score);
    Console.WriteLine($"Answer: {answer}, Score: {score}");
}
```

Also run existing unit tests for classification, NER, embedding, and reranking to confirm no regressions.

### Environment

- Model: `lquint/minilm-uncased-squad2-onnx` (BertForQuestionAnswering, hidden_size=384)
- Exported from `deepset/minilm-uncased-squad2` via `optimum-cli` (PR #24)
- ONNX Runtime: 1.21.x
- Discovered via end-to-end testing in [model-packages-prototype](https://github.com/luisquintanilla/model-packages-prototype)

### PR DESCRIPTION
**Pull request by luisquintanilla**
## Summary

Fixes `DiscoverModelMetadata()` throwing `InvalidOperationException("Could not determine hidden dimension from ONNX output 'start_logits'")` when used with QA models whose output tensors have all-dynamic dimensions.

Fixes #25

## Root Cause

QA models (e.g. `minilm-uncased-squad2`) output `start_logits` and `end_logits` with shape `[batch_size, sequence_length]` where **both** dimensions are dynamic (`[-1, -1]`). The `hiddenDim = (int)dims.Last()` extraction yields `-1`, which triggers the guard clause.

The adjacent code for `AdditionalOutputTensorNames` already handled this correctly by falling back to `MaxTokenLength` when the last dimension is dynamic.

## Fix

Apply the same `MaxTokenLength` fallback after every `hiddenDim = (int)dims.Last()` assignment  all 5 branches in `DiscoverModelMetadata()`:

1. `OutputTensorName` explicit branch
2. `PreferredOutputNames` branch (preferred found)
3. `PreferredOutputNames` branch (preferred not found, fallback)
4. Auto-discovery pooled branch (`sentence_embedding` / `pooler_output`)
5. Auto-discovery unpooled branch (`last_hidden_state` / `output`)

The existing guard clause (`hiddenDim <= 0  throw`) is kept as a safety net for the edge case where `MaxTokenLength` is also `<= 0`.

## Impact

- **QA models**: No longer throws. `hiddenDim` resolves to `MaxTokenLength` (128 by default).
- **Embedding models**: Hidden dim is static (e.g., 384). Fallback never triggers. No behavior change.
- **Classification/NER/Reranking models**: 3D output tensors with static last dim. Fallback never triggers. No behavior change.

## Changes

`src/MLNet.TextInference.Onnx/OnnxTextModelScorerEstimator.cs`  10 lines added (2 per branch  5 branches)

### CODE REVIEW

File: src/MLNet.TextInference.Onnx/OnnxTextModelScorerEstimator.cs [hunk 1]
```diff
@@ -225,6 +233,8 @@ internal OnnxModelMetadata DiscoverModelMetadata(InferenceSession session)
                     outputMeta.Keys.First());
                 hasPooledOutput = false;
                 hiddenDim = (int)outputMeta[outputName].Dimensions.Last();
+                if (hiddenDim <= 0)
+                    hiddenDim = _options.MaxTokenLength;
```
[the remark this answers is not in the document]
Comment (luisquintanilla) [reply src/MLNet.TextInference.Onnx/OnnxTextModelScorerEstimator.cs:237]: **Deferred.** You're right  for rank-3 outputs (\last_hidden_state\) the last dim is hidden_dim, not sequence_length. Falling back to \MaxTokenLength\ is semantically wrong. In practice this never triggers (no model has a dynamic hidden dimension), but the code should throw here instead of silently using a wrong value. Tracked in #28 (Issue 2).

File: src/MLNet.TextInference.Onnx/OnnxTextModelScorerEstimator.cs [hunk 2]
```diff
@@ -186,6 +186,8 @@ internal OnnxModelMetadata DiscoverModelMetadata(InferenceSession session)
             var dims = outputMeta[outputName].Dimensions;
             hasPooledOutput = !dims.Contains(-1) && dims.Length == 2;
             hiddenDim = (int)dims.Last();
+            if (hiddenDim <= 0)
+                hiddenDim = _options.MaxTokenLength;
```
[the remark this answers is not in the document]
Comment (luisquintanilla) [reply src/MLNet.TextInference.Onnx/OnnxTextModelScorerEstimator.cs:190]: **Deferred.** Good suggestion  the 2-line pattern is repeated 5 times. A \ResolveLastDimension(dims, tensorName, allowDynamicFallback)\ helper would centralize the logic and combine deduplication with the semantic fix for pooled/rank-3 branches. Tracked in #28 (Issue 3).

File: src/MLNet.TextInference.Onnx/OnnxTextModelScorerEstimator.cs [hunk 3]
```diff
@@ -197,6 +199,8 @@ internal OnnxModelMetadata DiscoverModelMetadata(InferenceSession session)
                 var dims = outputMeta[preferred].Dimensions;
                 hasPooledOutput = !dims.Contains(-1) && dims.Length == 2;
                 hiddenDim = (int)dims.Last();
+                if (hiddenDim <= 0)
+                    hiddenDim = _options.MaxTokenLength;
```
[the remark this answers is not in the document]
Comment (luisquintanilla) [reply src/MLNet.TextInference.Onnx/OnnxTextModelScorerEstimator.cs:203]: **Deferred  pre-existing bug, not introduced by this PR.** The \!dims.Contains(-1) && dims.Length == 2\ check was here before PR #26. You're right that \[-1, 384]\ (dynamic batch, static hidden) gets \hasPooledOutput = false\ when it should be \	rue\. The fix is \hasPooledOutput = dims.Length == 2 && dims.Last() > 0\  batch being dynamic is normal and irrelevant to whether the output is pre-pooled. Tracked in #28 (Issue 1).

File: src/MLNet.TextInference.Onnx/OnnxTextModelScorerEstimator.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (luisquintanilla) [reply src/MLNet.TextInference.Onnx/OnnxTextModelScorerEstimator.cs:190]: **Deferred  same pre-existing bug as the PreferredOutputNames comment above.** The \!dims.Contains(-1)\ check rejects dynamic batch dimensions, which is wrong. Same fix: \hasPooledOutput = dims.Length == 2 && dims.Last() > 0\. Tracked in #28 (Issue 1).

File: src/MLNet.TextInference.Onnx/OnnxTextModelScorerEstimator.cs [hunk 4]
```diff
@@ -216,6 +222,8 @@ internal OnnxModelMetadata DiscoverModelMetadata(InferenceSession session)
                 outputName = pooledName;
                 hasPooledOutput = true;
                 hiddenDim = (int)outputMeta[pooledName].Dimensions.Last();
+                if (hiddenDim <= 0)
+                    hiddenDim = _options.MaxTokenLength;
```
[the remark this answers is not in the document]
Comment (luisquintanilla) [reply src/MLNet.TextInference.Onnx/OnnxTextModelScorerEstimator.cs:226]: **Deferred.** Same class of issue as the rank-3 comment  \MaxTokenLength\ is the wrong fallback for pooled embedding outputs where the last dim is hidden_dim (e.g., 384), not sequence_length. No model has dynamic hidden dim for \sentence_embedding\ so this never triggers today, but the code should throw instead of silently using a wrong value. Your suggested error message is good. Tracked in #28 (Issue 2).
